### ISSUE TITLE
Generated badge with port version

### ISSUE THREAD

**Issue by AMDmi3** [FreshPorts/freshports#3]
Just a crazy idea. You've likely seen these [badges](http://shields.io/#version) used by many GitHub repos to report continous integration status and other stuff. Why not make FreshPorts generate a badge for a port by request? The badge should simply show DISTVERSION, like this:

![FreeBSD port](https://img.shields.io/badge/Freshports-7.0.27-blue.svg)
![FreeBSD port](https://img.shields.io/badge/FreeBSD_port-7.0.27-blue.svg)

This has numerous purposes among which is raising awareness of FreeBSD ports/Freshports among users, a convenient way to know the project is supported by FreeBSD and a way to know that the port is at latest version (or whether it falls behind).

The simplest way to implement this is just a make a per-port link (e.g. http://www.freshports.org/x11/xproto/badge or http://www.freshports.org/x11/xproto/?badge) which redirects to shields.io like ```https://img.shields.io/badge/FreeBSD_port-${DISTVERSION}-blue.svg```. Or you could just generate the same .svg by yourself to not depend on third party site.

**Comment by dlangille** [comment]
I do not recall seeing badges before.

Where would these badges appear?

**Comment by AMDmi3** [comment]
For example, there's pypi badge in ansible you've forked:

https://github.com/dlangille/ansible#readme

Actually they are used in most repos:

https://github.com/chef/chef#readme
https://github.com/bower/bower#readme
https://github.com/docker/docker#readme

**Comment by dlangille** [comment]
I see what you mean now.  It is a way for a given project to indicate that their software is available on FreeBSD.

**Comment by AMDmi3** [comment]
Yes, however if you put it that way a project could just generate a badge themselves. With Freshports though, the badge will also show version which is a lot more useful.

**Comment by dlangille** [comment]
Yes, the above is what I thought you were getting at.  I suspect `badges.freshports.org/x11/xproto` sounds like fun.

**Comment by dlangille** [comment]
I'd like `commits.freshports.org` instead of `commit.php?message_id....`

The goal being to make `www.freshports.org` 'cleaner'.

**Comment by AMDmi3** [comment]
Yes, would be great!

You could also set up a script to generate markdown code suitable for inclusion in README.md, e.g.

```
[![FreeBSD port](http://badges.freshports.org/${CATEGORY}/${PORT})](http://freshports.org/${CATEGORY}/${PORT})
```

(this is badge image which links to port's page on freshports) and add link to that from

```SVNWeb : Homepage : Distfiles Availability : PortsMon : Badge```

line on a ports page.

**Comment by dlangille** [comment]
After a short discussion at FreeBSD Developers summit in Stockholm, I think `www.freshports.org/--/badges/` may be a better idea.  Fewer certificates, host names, and web instances to maintain.  It could also be used `www.freshports.org/--/API/` (re https://duck.co/ideas/idea/936/grab-bsd-port-summary-from-freshports)

**Comment by AMDmi3** [comment]
I'm good with either. Also :+1: to API, as with it FP could be also ingerdated with shields.io directly.

**Comment by dlangille** [comment]
See http://dev.freshports.org/--/badges/?port=x11/xproto

**Comment by dlangille** [comment]
Proof of concept.  There is no error handling yet (always returns 200).

**Comment by AMDmi3** [comment]
Well not quite. It should either generate a redirect to badges.io link which it self shows a needed svg (example of such link is in the top post), or json with enough info for badges.io to generate it itself (which would be even better).

**Comment by AMDmi3** [comment]
Actually badge can be generated from this as well, but i'd name it ```/version/``` instead of ```/badge/``` then, and JSON would still be more flexible.

**Comment by AMDmi3** [comment]
Oops, accidental close.

**Comment by dlangille** [comment]
FYI badges.io is for sale.

**Comment by dlangille** [comment]
I got it: https://img.shields.io/badge/Bacula--Server-7.0.2_1-green.svg?style=plastic

**Comment by dlangille** [comment]
Examples:

https://img.shields.io/badge/Bacula--Server-7.0.2_1-red.svg?style=plastic
https://img.shields.io/badge/bacula--server-7.0.2__1-red.svg

Next step: figure out JSON, which I'm not seeing a reference to on their website.

**Comment by dlangille** [comment]
At present, http://dev.freshports.org/--/badges/?port=sysutils/bacula-client produces an IMG element if found, and a 404 otherwise.

**Comment by AMDmi3** [comment]
- It should produce link instead of  img element as the latter cannot be embedded into markdown.
- Version shouldn't include PORTREVISION and PORTEPOCH
- Color should be blue by default (most packaging systems do so), and red if BROKEN/IGNORE is present
- Instead of 404, it should produce a link to gray badge with "not found" text or something similar

**Comment by AMDmi3** [comment]
- Left-hand text should say "FreeBSD port" so it's apparent what the badge represents
- This text should stay for `not found' case, e.g. **FreeBSD port | not found**

**Comment by dlangille** [comment]
Compare http://dev.freshports.org/--/badges/?port=archivers/hlextract and http://dev.freshports.org/--/badges/?port=sysutils/bacula-client

The img is included with the HTML for debugging.

**Comment by AMDmi3** [comment]
Apart from 2 items from my last comment, good.

**Comment by dlangille** [comment]
Try now, not sure yet about the Not Found issue: http://dev.freshports.org/--/badges/?port=not/foundtest

**Comment by AMDmi3** [comment]
Yes, almost good - I'd use lightgray instead as it seems pretty common for other badges.

Also note that for not found case generated URL != generated img src (though img src is correct).

**Comment by dlangille** [comment]
How's that?

**Comment by dlangille** [comment]
Still to do: caching.

**Comment by AMDmi3** [comment]
Absolutely perfect! Thank you!

**Comment by AMDmi3** [comment]
Could you please remove <img> now? I'd like to test this on one of my repos.

**Comment by dlangille** [comment]
Try now.   Before moving to production, I suspect the URL / parameters may change.

**Comment by AMDmi3** [comment]
Uh, it should redirect (with 303 code I think) to the url, not show it.

**Comment by dlangille** [comment]
Done

**Comment by AMDmi3** [comment]
Yeah, now it works:

https://github.com/AMDmi3/libSDL2pp/tree/ports-badge

Waiting for production state of this feature :)

**Comment by dlangille** [comment]
I changed to lower case.  Seems they are caching it.

**Comment by AMDmi3** [comment]
Yep, github could. Direct link has lowercase `port`.

**Comment by AMDmi3** [comment]
Although it is disturbing, seems like it's caching the redirect while it shouldn't - this means that even if the version is updated, GH will still show an old badge. According ti http://www.w3.org/Protocols/rfc2616/rfc2616-sec10.html, you should redirect with 303 (which is non cacheable) or provide corresponding Cache-Control / Expires headers.

**Comment by dlangille** [comment]
Code was: ```header("Location: $badge");```

now: header("Location: $badge", true, 303);

re http://php.net/manual/en/function.header.php

**Comment by AMDmi3** [comment]
Thank you! We'll see if it helps.

**Comment by dlangille** [comment]
FYI, I am creating a branch, badges, so I can commit this work in progress without affecting production.  Then I will checkout this code for use on http://beta.freshports.org/ and that will allow me to keep working on dev, which is my personal sandbox.

**Comment by dlangille** [comment]
This should now work on http://beta.freshports.org/ : 

* http://dev.freshports.org/--/badges/?port=archivers/hlextract
* http://dev.freshports.org/--/badges/?port=sysutils/bacula-client
* http://dev.freshports.org/--/badges/?port=fake/fake

**Comment by AMDmi3** [comment]
> Please update your URLs to point to beta, but anticipate that the URL / form may change after REST is implemented.

Done

**Comment by dlangille** [comment]
I just pushed the badges branch.  It's no longer just on my dev site.

**Comment by AMDmi3** [comment]
E.g. it should be available on freshports.org? It doesn't seem to (or is the url different?):

```
 Sorry, but I don't know anything about that.

/--/badges/?port=archivers/hlextract
```

**Comment by dlangille** [comment]
No, not yet.  It is on the branch and needs to be merged into the main tree, but I want to configure caching before I do that.

**Comment by AMDmi3** [comment]
Any news on this?

**Comment by dlangille** [comment]
It is sad isn't it?

I'm now busy with EuroBSDCon prep...

**Comment by dlangille** [comment]
I managed to get this working in VMware on my laptop.  I'm flying tomorrow.

**Comment by dlangille** [comment]
I wonder what I need to cache here.  It is just a redirect.

Here's the fetch:

> [dan@dent:~] $ wget -S http://192.168.165.128/--/badges/?port=archivers/hlextract
--2016-10-21 21:42:18--  http://192.168.165.128/--/badges/?port=archivers/hlextract
Connecting to 192.168.165.128:80... connected.
HTTP request sent, awaiting response... 
  HTTP/1.1 303 See Other
  Date: Thu, 20 Oct 2016 00:42:55 GMT
  Server: Apache/2.2.31 (FreeBSD) PHP/5.5.32 mod_ssl/2.2.31 OpenSSL/1.0.2f DAV/2
  X-Powered-By: PHP/5.5.32
  Location: https://img.shields.io/badge/FreeBSD_port-2.4.4-red.png
  Content-Length: 0
  Keep-Alive: timeout=5, max=100
  Connection: Keep-Alive
  Content-Type: text/html
Location: https://img.shields.io/badge/FreeBSD_port-2.4.4-red.png [following]
--2016-10-21 21:42:18--  https://img.shields.io/badge/FreeBSD_port-2.4.4-red.png
Resolving img.shields.io (img.shields.io)... 104.18.44.160, 104.18.45.160, 2400:cb00:2048:1::6812:2da0, ...
Connecting to img.shields.io (img.shields.io)|104.18.44.160|:443... connected.
HTTP request sent, awaiting response... 
  HTTP/1.1 200 OK
  Date: Sat, 22 Oct 2016 02:42:20 GMT
  Content-Type: image/png
  Transfer-Encoding: chunked
  Connection: keep-alive
  Set-Cookie: __cfduid=d47352fea27e152c5a71682f8bfc382501477104140; expires=Sun, 22-Oct-17 02:42:20 GMT; path=/; domain=.shields.io; HttpOnly
  Cache-Control: max-age=86400
  Last-Modified: Sat, 22 Oct 2016 00:29:12 GMT
  Server: cloudflare-nginx
  CF-RAY: 2f59986f8bb358af-DFW
Length: unspecified [image/png]
Saving to: ‘index.html?port=archivers%2Fhlextract’

index.html?port=archivers%2Fhlextract            [ <=>                                                                                           ]   2.42K  --.-KB/s    in 0.001s  

2016-10-21 21:42:21 (4.27 MB/s) - ‘index.html?port=archivers%2Fhlextract’ saved [2478]

[dan@dent:~]

**Comment by dlangille** [comment]
Everything is just read from the DB.  That should be fast for now.  I'll see what I need to do to get this working on the plane tomorrow,

**Comment by dlangille** [comment]
I do believe this has been merged from the badges branch to the master branch.

https://github.com/FreshPorts/freshports/commit/4e202da5ffe22e4c52c9b9d86280ab924f41494a

I'll cleanup the files during the flight and this should be on dev.freshports.org later this weekend.

**Comment by dlangille** [comment]
Reopening, until we are in production.

**Comment by dlangille** [comment]
I see we are in production.  I missed that.

How's it looking?

**Comment by dlangille** [comment]
Still waiting to hear back from OP.

badges question raised again here: https://twitter.com/f0andrey/status/1049027503199858688

**Comment by grahamperrin** [comment]
https://github.com/FreshPorts/freshports/issues/3#issuecomment-150213603

> … for 'not found' case, …

See also: 

* #384

Is there any overlap? (If not, I can delete this comment.)
